presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.re2j.Matcher;
+import com.google.re2j.Options;
+import com.google.re2j.Pattern;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.primitives.Ints.checkedCast;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class Re2JRegexp
+{
+    private static final Logger log = Logger.get(Re2JRegexp.class);
+
+    private static final java.util.regex.Pattern DOT_STAR_PREFIX_SUFFIX_PATTERN = java.util.regex.Pattern.compile("^(\\.\\*\\??)?(.*?)(\\.\\*\\??)?$");
+    private static final int CORE_PATTERN_INDEX = 2;
+
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public final Pattern re2jPattern;
+    public final Pattern re2jPatternWithoutDotStartPrefixSuffix;
+
+    public Re2JRegexp(int dfaStatesLimit, int dfaRetries, Slice pattern)
+    {
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+
+        Options options = Options.builder()
+                .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                .setNumberOfDFARetries(dfaRetries)
+                .setEventsListener(new RE2JEventsListener())
+                .build();
+
+        String patternString = pattern.toStringUtf8();
+        re2jPattern = Pattern.compile(patternString, options);
+
+        // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+        java.util.regex.Matcher dotStarPrefixSuffixMatcher = DOT_STAR_PREFIX_SUFFIX_PATTERN.matcher(patternString);
+        checkState(dotStarPrefixSuffixMatcher.matches());
+        String patternStringWithoutDotStartPrefixSuffix = dotStarPrefixSuffixMatcher.group(CORE_PATTERN_INDEX);
+
+        if (!patternStringWithoutDotStartPrefixSuffix.equals(patternString)) {
+            re2jPatternWithoutDotStartPrefixSuffix = Pattern.compile(patternStringWithoutDotStartPrefixSuffix, options);
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
```

Comment:
Let me put some context into this optimization. In both RE2 and RE2J-TD there is an optimization in DFA that quickly searches for first byte of a potential pattern match. First byte(s) are determined by looking for bytes which make state machine to leave DFA starting state. When doing `find` the DFA uses special unanchored version of program. This is basically `any_byte*pattern` When `pattern` is additionally prefixed with `.*` then the optimization doesn't work because every byte is potentially a beginning of a match. This applies to both C++ RE2 and RE2-TD.

`re2jPatternWithoutDotStartPrefixSuffix` optimizes the case when `find` is performed  on `.*pattern.*` AND we ignore match begin. It doesn't work when we care about match beginning. Keep in mind that using `find` with `.*pattern.*` doesn't make much sense, because number of potential matches in quadratic. As far as I know C++ RE2 doesn't address this issue either and simply executes `find` using DFA on `.*pattern.*`.

Now, it would be possible to handle this singular case internally in RE2J-TD. However, such change would be intrusive and cause unnecessary overhead (e.g. two DFA/NFA state machines for legitimate `.*patterns`, two compiles, combined handling of DFA retries and number of states, etc). Note that `.*pattern.*` might be used with `match` method, where DFA first byte optimization works correctly.

Because of that, I think it's better for now to keep `re2jPatternWithoutDotStartPrefixSuffix` optimization here. It is meant to optimize user suboptimal queries. RE2 users (programmers) wouldn't normally use `find` with `.*pattern.*`. If you care about moving the optimization to RE2J itself, I would prefer to try to add it with other RE2J optimizations. There are still some optimizations from C++ RE2 (OnePass engine, SearchBitState, done states) that are worth porting to RE2J-TD.

RE2 and Regex is a tricky stuff. If you have some additional questions, please let me know.